### ISSUE TITLE
Local node stops working when uploading larger file

### ISSUE THREAD

**Issue by zbiljic** [ethersphere/bee#894]
#### Summary
When uploading larger file (~300MiB), at one point node started only logging following error, two on one second intervals:

```
DEBU[2020-11-02T15:07:40+01:00] localstore push subscription iteration: index iterator function: get value: leveldb: not found
DEBU[2020-11-02T15:07:40+01:00] localstore push subscription iteration: index iterator function: get value: leveldb: not found
```

The thing that may relate to this problem is low configuration for `db-capacity`, which was `10000`.

#### Steps to reproduce
The node was connected to the `staging` network, with following configuration:

```yaml
data-dir: ~/tmp/bee/node
db-capacity: 10000
password: pass
debug-api-enable: true
tracing-enable: true
verbosity: trace
network-id: 2
bootnode:
- /dnsaddr/bootnode.staging.ethswarm.org
global-pinning-enable: false
swap-endpoint: https://rpc.slock.it/goerli
swap-enable: false
```

The file was being uploaded via `files` API.

The files was being uploaded with a tag, and this is the state at which upload stopped:

```json
{
    "address": "",
    "name": "unnamed_tag_1604325530",
    "seen": 1,
    "sent": 1674,
    "split": 31167,
    "startedAt": "2020-11-02T14:58:50+01:00",
    "stored": 31166,
    "synced": 1674,
    "total": 0,
    "uid": 2044808445
}
```

#### Expected behavior
The node should accept the upload, and push all chunks for the file.

#### Actual behavior
The node logged same 2 log lines every second, and upload was blocked.

@janos

**Comment by janos** [comment]
This is a bug of inconsistent state in localstore. The localstore itself is allowing this to happen and this should be fixed.

There are currently two ways how a chunk can enter the garbage collection process, one with ModeSetSyncPush and another ModeSetSyncPull. Both are adding the chunk to the garbage collection, but only ModeSetSyncPush is removing the chunk from the push sync index. If ModeSetSyncPull happens before ModeSetSyncPush, the chunk can be garbage collected before the pusher reaches to its address in push sync index, resulting that the chunk is not found on every iteration.

Localstore should not allow to enter in inconsistent state by using its api, and to relay that other components would not use it in a way that it can create an inconsistent state.

There should be only one ModeSetSync which will mark the chunk as synced and put it into the garbage collection index. When that should happen is open for discussion and should be clearly defined, so that we know when the chunk can be garbage collected safely.

**Comment by Eknir** [comment]
Need input from research @zelig

**Comment by Eknir** [comment]
@janos , dragged this to `In Progress` in the active Sprint workspace. Thanks for correctly attaching the PR to the issue :thumbsup:

**Comment by janos** [comment]
Thanks @Eknir. The PR is ready, but would not merge it before we have a fix for gc integration test.

**Comment by Eknir** [comment]
@zbiljic, could you please verify that @janos PR fixes the problem (and move the issue in the QA pipeline)

**Comment by zbiljic** [comment]
I have started and checked current master build (`a3fdf8bc92a08a1689b8542c728db3d4e98ac5b7`), and it no longer exhibits the issue reported here.

Now some other (but maybe expected) problems were shown. Since the configuration that was use for this is set to really low capacity, there was some strange behavior.
Uploading ~300MB file succeeds (HTTP request completes), however after that since the node did not have any credit, it tried to settle with peers in a loop. The CPU usage was a bit higher than expected, the network throughput was downloading ~4Mb/s, and only uploading ~100Kb/s (this is on 500/20 Mb/s). I assume node was trying to take on other chunks in order to have more balance so it could use it for its own chunks. This continued for some time ~30min, after which I restarted once (to check how it behaves), and after that terminated node. Looking at tags it did seem like the node was sending chunks (before stopping it ended at 12092/73008).
The file that was uploaded was not available after upload.

I would say the issue in question no longer occurs.
